html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php

```diff
@@ -0,0 +1,337 @@
+    return 'node';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getTitle() {
+    return $this->t('My posts');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getStatuses() {
+    return [
+      'draft' => $this->t('draft'),
+      'pending' => $this->t('pending'),
+      'published' => $this->t('published'),
+      'on_hold' => $this->t('on-hold'),
+      'refused' => $this->t('refused'),
+      'expired' => $this->t('expired'),
+      'duplicate' => $this->t('duplicate'),
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getHeaders() {
+    return [
+      'id' => [
+        'label' => $this->t('Id'),
+        'type' => 'property',
+        'specifier' => 'nid',
+        'sortable' => TRUE,
+      ],
+      'type' => [
+        'label' => $this->t('Type'),
+        'type' => 'property',
+        'specifier' => 'type',
+        'sortable' => TRUE,
+      ],
+      'status' => [
+        'label' => $this->t('Status'),
+        'type' => '',
+        'specifier' => 'moderation_state',
+        'sortable' => TRUE,
+      ],
+      'poster' => [
+        'label' => $this->t('Poster'),
+      ],
+      'source' => [
+        'label' => $this->t('Source'),
+      ],
+      'title' => [
+        'label' => $this->t('Title'),
+        'type' => 'property',
+        'specifier' => 'title',
+        'sortable' => TRUE,
+      ],
+      'date' => [
+        'label' => $this->t('Posted'),
+        'type' => 'property',
+        'specifier' => 'created',
+        'sortable' => TRUE,
+      ],
+      'deadline' => [
+        'label' => $this->t('Deadline'),
+        'type' => 'custom',
+        'specifier' => 'deadline',
+        'sortable' => TRUE,
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRows(array $results) {
+    if (empty($results['entities'])) {
+      return [];
+    }
+
+    /** @var \Drupal\reliefweb_moderation\EntityModeratedInterface[] $entities */
+    $entities = $results['entities'];
+
+    // Prepare the table rows' data from the entities.
+    $rows = [];
+    foreach ($entities as $entity) {
+      $cells = [];
+
+      // Edit link + status cell.
+      $cells['id'] = $entity->id();
+      $cells['type'] = $entity->bundle();
+      $cells['status'] = new FormattableMarkup('<div class="rw-moderation-status" data-moderation-status="@status">@label</div>', [
+        '@status' => $entity->getModerationStatus(),
+        '@label' => $entity->getModerationStatusLabel(),
+      ]);
+      $cells['poster'] = $this->currentUser->id() == $entity->getOwner()->id() ? $this->t('me') : $this->t('other');
+      $cells['source'] = '';
+      $cells['title'] = $entity->toLink()->toString();
+      $cells['date'] = $this->getEntityCreationDate($entity);
+
+      $cells['deadline'] = '';
+      if ($entity->field_registration_deadline) {
+        $cells['deadline'] = $this->formatDate($entity->field_registration_deadline->value);
+      }
```

Comment:
Note: it's allowed for a training to have an empty `field_registration_deadline`: for **ongoing** training. In that case the deadline label should be set to "`$cells['deadline'] = $this->t('Ongoing')`". So probably need to check the entity bundle in addition to the field value.

Reply:
Fixed, but ugly, for some reason `$entity->field_registration_deadline->value` returns the string `Array`, so `IsEmpty()` does not work

Reply:
@attiks Would you have the ID of a node for which this happen?

Reply:
I just created an ongoing training and opened `/moderation/content/training`